### ISSUE TITLE
Is the data augmentation performed accurately?

### ISSUE THREAD

**Issue by trangnambeo** [petrosgk/Kaggle-Carvana-Image-Masking-Challenge#13]
Hi,

Is the following code in train.py correct?

for id in ids_train_batch.values:
                img = cv2.imread('input/train/{}.jpg'.format(id))
                img = cv2.resize(img, (input_size, input_size))
                mask = cv2.imread('input/train_masks/{}_mask.png'.format(id), cv2.IMREAD_GRAYSCALE)
                mask = cv2.resize(mask, (input_size, input_size))
                img = randomHueSaturationValue(img,
                                               hue_shift_limit=(-50, 50),
                                               sat_shift_limit=(-5, 5),
                                               val_shift_limit=(-15, 15))
                img, mask = randomShiftScaleRotate(img, mask,
                                                   shift_limit=(-0.0625, 0.0625),
                                                   scale_limit=(-0.1, 0.1),
                                                   rotate_limit=(-0, 0))
                img, mask = randomHorizontalFlip(img, mask)
                mask = np.expand_dims(mask, axis=2)
                x_batch.append(img)
                y_batch.append(mask)

From what I understood so far about data augmentation, the idea is "adding" more training data but the code above "update" original image/mask, apply image processing algorithms on them in a series manner then use the final processed ones as training data. So, basically, we did not increase training data. Should we perform the following tasks instead?
- Add original img/mask to x_batch/y_batch
- Add every output of data augmentation algorithm to x_batch/y_batch

I know that It will increase training time a lot but looks like it is the right thing to do. Please correct me If I'm wrong.

Thanks,

**Comment by killthekitten** [comment]
The augmentation is performed on a random basis with a `0.5` threshold. That actually means that on average, each image will be augmented only in half of the training epochs, being slightly changed each time. So the number of augmentations will be big, keeping the single epoch training time the same. Most likely, every image from the original set will be shown to the network at least once.

Also, you can tune it to get a better performance. For example, you can decay the threshold each epoch in such a way, that eventually only originals will be fed to the network.

**Comment by trangnambeo** [comment]
Got it, now I understand the meaning of random augmentation. Thank you guys alot.
